### PR TITLE
add incremental_settings design doc

### PR DESCRIPTION
**Pull request by richm**
design doc for incremental settings

### CODE REVIEW

File: documentation/incremental_settings.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  - name: fs.file-max
+    state: absent
+  - name: kernel.threads-max
+    state: absent
+```
+This will add/modify the parameter `fs.epoll.max_user_watches` in the managed
+object (in this case the managed object is the OS `sysctl` settings), and
+remove the the parameters `fs.file-max` and `kernel.threads-max` (in this
+case, this means the values will revert back to their system default values,
+usually upon system reboot).
```
Comment (pcahyna) [inline documentation/incremental_settings.md:47]: a note specific to this `kernel_settings` example and not to the general concept: to me it does not seem entirely right to advertise this capability (of removing values) because the result is not well defined:
- while other changes are applied immediately, this one takes effect after reboot, which is not consistent (of course, changes of bootloader parameters also require reboot, but there it is obvious, while touching a parameter that can be configured in runtime is expected to be visible immediately)
- `absent` does not really cause the underlying object (the entry in the sysctl MIB tree) to be absent. It rather sets it to some (unknown) default. This is inconsistent with the usual meaning of `absent` and also it is not clear what the result is.

The second point depends of what you actually consider the role to be managing. If you consider it a role to manage tuned profiles, then your approach makes complete sense: indeed, `state: absent` will make sure that the setting has disappeared from the tuned profile, as expected. But if it is a role that manages kernel settings, and tuned is merely a provider (implementation detail), then the point is relevant, because the role can't actually remove the settings.
In other words, `state: absent` is breaking abstraction. It is a concept valid for the provider but not so much for the general role.

File: documentation/incremental_settings.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+This will remove any existing settings and replace them with the settings given.
+
+## Removing all of the settings with state: empty
+
+This is the case where you want to remove all of the existing values in the
+managed object.  For example, if you want to remove all of the
+`kernel_settings_sysctl` parameters:
+```yaml
+kernel_settings_sysctl:
+  state: empty
```
Comment (pcahyna) [inline documentation/incremental_settings.md:74]: It could be mentioned that this is exactly equivalent, but more readable, version of
```
kernel_settings_sysctl:
  - previous: replaced
```
as it specifies a list that is empty except the `previous: replaced` special value which drops the previous settings.

File: documentation/incremental_settings.md (hunk 1 above)
Comment (pcahyna) [reply documentation/incremental_settings.md:47]: I would keep `state: absent` in the kernel_settings role, but choose a different example here to illustrate the general concept.

File: documentation/incremental_settings.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# Incremental Settings for Ansible Values
+
+The general approach in Ansible is that you provide all of the values for the
```
Comment (pcahyna) [inline documentation/incremental_settings.md:3]: This is not entirely true. Ansible modules generally are using the concept outlined here (perform pointed changes and preserve everything else). The sysctl module does this and it seems to be the norm rather than an exception. That's why the interface of modules usually translates in a fairly straightforward way to the role interfaces described here. It is roles and playbooks that usually describe the complete state of the system they are managing.

File: documentation/incremental_settings.md (hunk 1 above)
Comment (richm) [reply documentation/incremental_settings.md]: addressed in https://github.com/linux-system-roles/linux-system-roles.github.io/pull/21/commits/4577d2c20cb2a14ed176486cb823442645d7088b

File: documentation/incremental_settings.md (hunk 2 above)
Comment (richm) [reply documentation/incremental_settings.md:82]: addressed in https://github.com/linux-system-roles/linux-system-roles.github.io/pull/21/commits/4577d2c20cb2a14ed176486cb823442645d7088b

File: documentation/incremental_settings.md (hunk 3 above)
Comment (richm) [reply documentation/incremental_settings.md]: addressed in https://github.com/linux-system-roles/linux-system-roles.github.io/commit/4577d2c20cb2a14ed176486cb823442645d7088b

File: documentation/incremental_settings.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
+# Incremental Settings for Ansible Values
+
+In some cases in Ansible, you provide most or all of the pertient attributes
+of the managed object.  For example, for the `file` module, you provide the
+content, permissions, ownership, and other attributes of the file to be
+managed. In other cases, you have to specify how to change the *incremental*
```
Comment (pcahyna) [inline documentation/incremental_settings.md:6]: I would write "how to *incrementally* change the state of the system" instead of "how to change the *incremental* state of the system". Incremental state to me sounds like an oxymoron (state is not the increment, but the thing on top of which you apply the increment and the result (another state). I.e. state is an integral, increment a derivative. (Actually a sum and a difference, but IKWYM)).

File: documentation/incremental_settings.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
+# Incremental Settings for Ansible Values
+
+In some cases in Ansible, you provide most or all of the pertient attributes
+of the managed object.  For example, for the `file` module, you provide the
+content, permissions, ownership, and other attributes of the file to be
```
Comment (pcahyna) [inline documentation/incremental_settings.md:5]: I think this is still misleading. The `file` module actually does very fine-grained, incremental changes just as `sysctl` does. Consider the following playbook, you will see that each invocation of `file` changes only one attribute and preserves the others.
```yaml
- name: show fine grained changes with the file module
  hosts: localhost
  vars:
    fname: foo
  tasks:
    - file:
        state: touch
        name: "{{ fname }}"
    - stat:
        path: "{{ fname }}"
      register: stat0
    - name: show initial result
      debug:
        msg: "{{ stat.gr_name }} {{ stat.mode }} {{ stat.mtime }}"
      vars:
        stat: "{{ stat0.stat }}"

    - file:
        mode: g+x
        name: "{{ fname }}"
    - stat:
        path: "{{ fname }}"
      register: stat1

    - name: show first result
      debug:
        msg: "{{ stat.gr_name }} {{ stat.mode }} {{ stat.mtime }}"
      vars:
        stat: "{{ stat1.stat }}"
    - file:
        group: libvirt
        name: "{{ fname }}"
    - stat:
        path: "{{ fname }}"
      register: stat2
    - name: show second result
      debug:
        msg: "{{ stat.gr_name }} {{ stat.mode }} {{ stat.mtime }}"
      vars:
        stat: "{{ stat2.stat }}"

    - file:
        modification_time: now
        name: "{{ fname }}"
    - stat:
        path: "{{ fname }}"
      register: stat3
    - name: show third result
      debug:
        msg: "{{ stat.gr_name }} {{ stat.mode }} {{ stat.mtime }}"
      vars:
        stat: "{{ stat3.stat }}"

    - file:
        state: absent
        name: "{{ fname }}"
```
If you need an example of an Ansible module that replaces the whole object, `template` or `copy` modules might be the best. I believe that if you go through the core modules, you will find though that incremental, fine-grained changes and preservation of unspecified values is actually the usual behavior for Ansible modules and replacing the whole managed object is more an exception.
I propose to change the wording as follows:
"Ansible playbooks and roles are often written so that they declare the complete state of the subsystem that they manage and entirely replace the previous state, such as the content of the managed configuration files. On the other hand, Ansible modules usually have an interface where you have to specify ..."

File: documentation/incremental_settings.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    name: some_setting
+    value: 123  # to add/replace the value
+```
+or
+```yaml
+- name: modify a single kernel setting
+  sysctl:
+    name: some_setting
+    state: absent  # to remove the value
+```
```
Comment (pcahyna) [inline documentation/incremental_settings.md:24]: I would change the first phrase to clarify the rationale why we are not following the usual pattern of Ansible roles and playbooks and rather follow the approach commonly taken by modules:
"Many roles in the System Roles project need to follow the usual pattern of module interfaces and allow specifying incremental changes instead of the complete state to be able to be safely used multiple times in a playbook without the later invocation clobbering the result of the previous invocation, and to preserve previous state of the managed systems. We had a need to generalize the usual approach of modules to lists of settings ..."
Alternatively, we can leave it as it is and add the rationale in a future commit.

**Comment by richm** [comment]
> I suppose you mean "_in_ a shorter, more readable version."
> I would provide the complete example here:
> 
> ```yaml
> widget_manager:
>   - previous: replaced
> ```
> 
> otherwise it may not be exactly clear where does the `- previous: replaced` item go.

fixed in https://github.com/linux-system-roles/linux-system-roles.github.io/pull/21/commits/373b5ac990e225b3ac11431fef1154a65e047475

File: documentation/incremental_settings.md (hunk 4 above)
Comment (richm) [reply documentation/incremental_settings.md]: Fixed in https://github.com/linux-system-roles/linux-system-roles.github.io/pull/21/commits/373b5ac990e225b3ac11431fef1154a65e047475

File: documentation/incremental_settings.md (hunk 5 above)
Comment (richm) [reply documentation/incremental_settings.md]: Fixed in https://github.com/linux-system-roles/linux-system-roles.github.io/pull/21/commits/373b5ac990e225b3ac11431fef1154a65e047475
Note that I got rid of the text to try to explain the difference between managing the whole state of the system or of the managed objects, and instead explained it as generalizing the concept of incremental changes as implemented by modules, and using that approach for system roles.

File: documentation/incremental_settings.md (hunk 6 above)
Comment (richm) [reply documentation/incremental_settings.md:21]: fixed in https://github.com/linux-system-roles/linux-system-roles.github.io/pull/21/commits/373b5ac990e225b3ac11431fef1154a65e047475 - I mostly used your text
